src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -97,7 +97,37 @@ public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
             if (dictionary.GetType() == typeof(Dictionary<TKey, TValue>))
             {
                 Dictionary<TKey, TValue> d = (Dictionary<TKey, TValue>)dictionary;
+
+                if (d._comparer == _comparer)
+                {
+                    // Comparers are the same, can copy dictionary wholesale
```

Comment:
Is it going to regress performance if somebody is using this method to compact the dictionary?

Reply:
Would be faster, but wouldn't compact. Should use the [`.TrimExcess()`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.trimexcess?view=netcore-3.1) method to compress it?

Alternative would be to create a new `.Clone()` extension/method #41939

Reply:
Or take this path only if there are no pending removed entries?

Reply:
Means you would go slow path if anything is deleted...

Changed to use the TrimExcess copy if there are free entries, so it also compresses on copy?

Reply:
Means you would go slow path if anything is deleted...

Changed to use the TrimExcess copy if there are free entries, so it also compresses on copy.

> Is it going to regress performance if somebody is using this method to compact the dictionary?

This way its a performance improvement if that is done also?

Reply:
Hmm, this can still regress performance if the original Dictionary has much larger capacity than what's necessary.

I am wondering whether it would be best to always rehash.

Reply:
I can change the condition to check what size the dictionary would be for count; then copy if same and trim if less

Reply:
It is likely going to be pretty rare to have the dictionary at the exact full capacity. It is what makes me think that it is best to always rehash.

Reply:
My mistake; should have been checking against `_entries.Length` not `.Count`.

i.e. if the prime for the count is going to generate the same sized arrays, then we can copy the arrays directly; if its a smaller value than the Length, then use smaller arrays and only copy over the filled entries.